### PR TITLE
Varvara/feature/reports search

### PR DESCRIPTION
**Pull request by varvarashcherbyna7**

### CODE REVIEW

File: app/models/visit.py [hunk 1]
```diff
@@ -23,3 +27,15 @@ class Visit(Base, ModelMixin):
...
+    @property
+    def visit_info(self):
+
+        return {
+            "id": self.id,
+            "date": self.date,
+            "start_time": self.start_time,
+            "end_time": self.end_time,
+            "client_info": self.client.client_info,
+            # "doctor": self.doctor,
```
Comment (bezryma) [inline app/models/visit.py:41]: we do not like commented code the so-called "for future" :)

File: app/routers/client.py [hunk 2]
```diff
@@ -105,3 +115,46 @@ async def client_intake(
 async def get_client_intake(api_key: str, doctor: Doctor = Depends(get_current_doctor)):
     """Returns client intake"""
     return ClientService.get_intake(api_key, doctor)
+
+
+# filtering for reports page
```
Comment (bezryma) [inline app/routers/client.py]: this comment dublicate the meaning of the endpoint`s docstring) so we can remove it

File: app/routers/client.py [hunk 3]
```diff
@@ -105,3 +115,46 @@ async def client_intake(
...
+@router_client.post(
+    "/report_visit", response_model=List[VisitReportRes], tags=["Client"]
+)
+async def formed_report_visit(
+    client_data: VisitReportReq, doctor: Doctor = Depends(get_current_doctor)
+):
+    """Filter for page reports visits by date"""
+    service = ReportService()
+    report_of_visit = service.filter_data_for_report_of_visit(client_data, doctor)
```
Comment (bezryma) [inline app/routers/client.py:128]: you can at once return the result of the service`s function call in this line without creating separate variable)

File: app/routers/client.py [hunk 4]
```diff
@@ -105,3 +115,46 @@ async def client_intake(
 async def get_client_intake(api_key: str, doctor: Doctor = Depends(get_current_doctor)):
     """Returns client intake"""
     return ClientService.get_intake(api_key, doctor)
+
+
+# filtering for reports page
+@router_client.post(
+    "/report_visit", response_model=List[VisitReportRes], tags=["Client"]
+)
+async def formed_report_visit(
```
Comment (bezryma) [inline app/routers/client.py]: I do not see that we "await" some corotine object in this functions... so does it should be async? I see you use dependency get_current_doctor which I suppose look for current user in db... Is all requests to the current db async?

File: app/routers/client.py [hunk 5]
```diff
@@ -105,3 +115,46 @@ async def client_intake(
...
+    service = ReportService()
+    report_of_visit = service.filter_data_for_report_of_visit(client_data, doctor)
+    return report_of_visit
+
+
+@router_client.get("/report_visit", response_class=FileResponse, tags=["Client"])
+async def report_visit(doctor: Doctor = Depends(get_current_doctor)):
+    """Get for page reports visits by date"""
+    file_report_path = "./visits_report.csv"
+    file = FileResponse(file_report_path)
```
Comment (bezryma) [inline app/routers/client.py]: again we can return at once FileResponse(file_report_path)

File: app/routers/client.py [hunk 6]
```diff
@@ -105,3 +115,46 @@ async def client_intake(
...
+    client_data: VisitReportReq, doctor: Doctor = Depends(get_current_doctor)
+):
+    """Filter for page reports visits by date"""
+    service = ReportService()
+    report_of_visit = service.filter_data_for_report_of_visit(client_data, doctor)
+    return report_of_visit
+
+
+@router_client.get("/report_visit", response_class=FileResponse, tags=["Client"])
+async def report_visit(doctor: Doctor = Depends(get_current_doctor)):
```
Comment (bezryma) [inline app/routers/client.py:133]: the same, should this function be async?

File: app/routers/client.py [hunk 7]
```diff
@@ -105,3 +115,46 @@ async def client_intake(
...
+
+@router_client.post(
+    "/report_new_clients", response_model=List[VisitReportResClients], tags=["Client"]
+)
+def formed_report_new_clients(
+    client_data: VisitReportReq, doctor: Doctor = Depends(get_current_doctor)
+):
+    """Filter for page reports new clients by date"""
+    service = ReportService()
+    report_of_new_clients = service.filter_data_for_report_of_new_clients(
```
Comment (bezryma) [inline app/routers/client.py:147]: again, we can return the result at once on this line

File: app/routers/client.py [hunk 8]
```diff
@@ -105,3 +115,46 @@ async def client_intake(
...
+        client_data, doctor
+    )
+    return report_of_new_clients
+
+
+@router_client.get("/report_new_clients", response_class=FileResponse, tags=["Client"])
+async def report_new_clients(doctor: Doctor = Depends(get_current_doctor)):
```
Comment (bezryma) [inline app/routers/client.py:154]: the same, should this function be async?

File: app/routers/client.py [hunk 9]
```diff
@@ -105,3 +115,46 @@ async def client_intake(
...
+    """Get for page reports visits by date"""
+    file_report_path = "./new_clients_report.csv"
+    file = FileResponse(file_report_path)
```
Comment (bezryma) [inline app/routers/client.py]: again, we can return the result at once on this line)

File: app/services/reports.py [hunk 10]
```diff
@@ -0,0 +1,207 @@
+from typing import List
+import csv
```
Comment (bezryma) [inline app/services/reports.py:3]: According to the CIA's internal coding conventions, python imports should be grouped into the following groups:
1. Standard library imports
2. Third-party imports
3. Application-specific imports
Imports should be ordered lexicographically within these groups, ignoring case. So according to this it is much better import first csv lib and then List from typing lib)

File: tests/test_client.py [hunk 11]
```diff
@@ -156,7 +155,7 @@ def test_get_queue(client: TestClient):
     data = {
         "api_key": client_intake.api_key,
         "rougue_mode": False,
-        "place_in_queue": client_intake.client_info["place_in_queue"],
+        # "place_in_queue": client_intake.client_info["place_in_queue"],
```
Comment (bezryma) [inline tests/test_client.py:158]: really need comment or we can remove it?)

File: tests/test_client.py [hunk 12]
```diff
@@ -184,7 +183,7 @@ def test_get_queue(client: TestClient):
     data = {
         "api_key": client_intake.api_key,
         "rougue_mode": True,
-        "place_in_queue": client_intake.client_info["place_in_queue"],
+        # "place_in_queue": client_intake.client_info["place_in_queue"],
```
Comment (bezryma) [inline tests/test_client.py:186]: really need comment or we can remove it?)

File: tests/test_client.py [hunk 13]
```diff
@@ -221,7 +220,7 @@ def test_get_queue(client: TestClient):
         "last_name": client_intake.last_name,
         "phone": client_intake.phone,
         "email": client_intake.email,
-        "place_in_queue": client_intake.client_info["place_in_queue"],
+        # "place_in_queue": client_intake.client_info["place_in_queue"],
```
Comment (bezryma) [inline tests/test_client.py:223]: really need comment or we can remove it?)

File: tests/test_client.py [hunk 14]
```diff
@@ -451,3 +450,63 @@ def test_complete_client_visit(client: TestClient):
...
+            end_time=datetime.datetime.strptime(time, "%m/%d/%Y, %H:%M:%S")
+            + datetime.timedelta(days=count + 2),
+            client_id=clientDB.id,
+            doctor_id=doctor.id,
+        ).save()
+        count = count + 1
+
+    visit: Visit = Visit.query.all()
+    assert visit
+    # assert len(visit) == 3
```
Comment (bezryma) [inline tests/test_client.py]: really need comment or we can remove it?)

File: tests/test_client.py [hunk 15]
```diff
@@ -451,3 +450,63 @@ def test_complete_client_visit(client: TestClient):
...
+    data = {
+        "type": "visit",
+        "start_time": visit1.start_time.strftime("%m/%d/%Y, %H:%M:%S"),
+        "end_time": visit3.end_time.strftime("%m/%d/%Y, %H:%M:%S"),
+    }
+    # 4. put date for filter report visits
+    response = client.post("/api/client/report_visit", json=data)
+    assert response
+    assert response.ok
+    # assert response.text
```
Comment (bezryma) [inline tests/test_client.py:485]: really need comment or we can remove it?)

**Review by bezryma — COMMENTED** [review]
Wonderful!

File: tests/test_client.py (hunk 15 above)
Comment (varvarashcherbyna7) [reply tests/test_client.py:485]: Thank you for all your comments, I will fix everything :)

File: app/routers/client.py (hunk 4 above)
Comment (varvarashcherbyna7) [reply app/routers/client.py]: I'm not sure here, you're right,) not all requests are asynchronous

File: app/routers/client.py (hunk 5 above)
Comment (varvarashcherbyna7) [reply app/routers/client.py]: corrected)

**Pull request merged**